added unicode catch to predict_proba function in SVC

JoshuaMeyers opened the pull request:
#### What does this implement/fix? Explain your changes.

Adds mixed object type handling (unicode/string) to the `kernel` parameter of SVM models.
This was present for the `predict` function, but neglected for `predict_proba`
For a recreation of the error, see https://github.com/EducationalTestingService/skll/issues/87

JoshuaMeyers committed:
added unicode catch to predict_proba function

See https://github.com/EducationalTestingService/skll/issues/87
Unicode handling in `.predict` had not been added to `.predict_proba`

rth commented:
Thank you for your PR.

Are you sure this wasn't fixed in https://github.com/scikit-learn/scikit-learn/pull/7064 ? If so please add a unit test (or extend [`test_unicode_kernel`](https://github.com/scikit-learn/scikit-learn/blob/ebcfcafcea873ef5067450bf548cd683ca9ab770/sklearn/svm/tests/test_svm.py#L507)) that would fail on master but pass with this PR. Thanks.

amueller commented:
you're mixing tabs and spaces. Please use spaces.
(also there's a weird issue in the test for tabs vs spaces?!)
Why is this necessary? It's not done in ``predict`` or in the sparse case. And in fitting it's only done in the dense case in ``_dense_fit``, right?

JoshuaMeyers committed:
Update base.py

JoshuaMeyers committed:
update unicode kernel unit test to include predict & predict_proba functions

Previously, unicode test only tested `clf.fit`.
In particular, `predict_proba` currently fails if unicode is provided to the `kernel` parameter

JoshuaMeyers committed:
added blank line

JoshuaMeyers commented:
Thanks for your responses!

@rth The code from the issue https://github.com/scikit-learn/scikit-learn/pull/7064 is the same but this was not applied to the `predict_proba` endpoint. I have also now added a call to `predict_proba` in the unit tests - this will fail on master.

@amueller My bad, I have fixed the spacing! It is necessary to catch unicode objects that are given to the `kernel` parameter which are not handled by `libSVM`. Other parameters are not sensitive to this and neither should `kernel` be. It is only necessary for the `_dense_predict_proba` also

JoshuaMeyers commented:
I'm having a nightmare with these checks haha! Why is LGTM failing? I don't understand the cause of the error. Help appreciated.

Commit:
added catch for kernel parameter if in ascii bytes in Python3

amueller commented on sklearn/svm/base.py:
```diff
@@ -322,6 +322,11 @@ def _dense_predict(self, X):
                                  "the number of samples at training time" %
                                  (X.shape[1], self.shape_fit_[0]))
 
+        if six.PY3:
```
Yeah I was confused as to why this didn't happen ;)

amueller commented on sklearn/svm/tests/test_svm.py:
```diff
@@ -505,27 +505,37 @@ def test_bad_input():
+        clf = svm.SVC(kernel=unicode('linear'), probability=True)
         clf.fit(X, Y)
+        clf.predict(T)
+        clf.predict_proba(T)
```
decision function?

JoshuaMeyers replied:
yep sounds a good idea, works on my local copy but just for completeness

amueller commented:
worry less about lgtm and more about travis ;)

Why is this not necessary in the sparse case?

Commit:
remove whitespace

Commit:
added clf.decision_function to unit test

JoshuaMeyers commented:
@amueller haha okay I'll focus on travis first.

As for why this isn't necessary in the sparse case, the kernel is specified by its index in the sparse case which seems not to be sensitive to this...
_sparse_predict: `kernel_type = self._sparse_kernels.index(kernel)`

I've tested and indeed, the sparse versions do not error.

amueller commented:
```python
    # The order of these must match the integer values in LibSVM.
    # XXX These are actually the same in the dense case. Need to factor
    # this out.
    _sparse_kernels = ["linear", "poly", "rbf", "sigmoid", "precomputed"]
```

Maybe we should just do that? It's weird to have substantially different code paths for sparse and dense. If you'd rather not get involved in this, that's fine, too ;)

amueller commented:
Basically we just need to move ``kernel_index = LIBSVM_KERNEL_TYPES.index(kernel)`` from the cython code to the python code in the dense case, then it'll be the same as the sparse case, we can rename ``_sparse_kernels`` to ``_kernels`` and can get rid of the string handling code and just pass integers to cython everywhere.

qmick commented:
It is still in progress?

qinhanmin2014 commented:
@qmick Feel free to take it if @JoshuaMeyers doesn't reply after some time. Note that we are using multiple ways to handle the problem currently.
(1)in sparse case : passing int kernel_type instead of str kernel
(2)in some functions in dense case(e.g., predict): passing kernel instead of str kernel
(3)in some functions in dense case(e.g., fit): using str to cast kernel to string before passing
I agree with amueller's comment above (i.e., always use the first way).

qmick commented:
@qinhanmin2014 Thanks, I'll keep looking at it.

JoshuaMeyers commented:
Hey @qmick, feel free to have a go at this. I haven't had the time. Apologies!

qinhanmin2014 commented:
Thanks @JoshuaMeyers for your great work so far :)

JoshuaMeyers commented:
This issue was addressed by https://github.com/scikit-learn/scikit-learn/pull/10412. Closing this MR.
